docs/references/architecture/adr-112-txi.md

```diff
@@ -0,0 +1,84 @@
+# ADR 112: Modular transaction hashing
+
+## Changelog
+
+- 2024-02-05: First version (@melekes)
+
+## Status
+
+TBD
+
+## Context
+
+Transaction hashing in CometBFT is currently implemented using `tmhash`
+package, which itself relies on `sha256` to calculate a transaction's hash.
+
+The hash is then used by the built-in indexer (to index this
+transaction) and by the RPC `tx` and `tx_search` endpoints, which allow users
+to search for a transaction using a hash.
+
+`tmhash` is also used to calculate various hashes in CometBFT (e.g.,
+evidence, consensus params, commit, header, partset header).
+
+The problem some application developers are facing is a mismatch between the
+internal/app representation of transactions and the one employed by CometBFT. For
+example, [Evmos](https://evmos.org/) wants transactions to be hashed using
+the [RLP][rlp].
+
+In order to be flexible, CometBFT needs to allow changing the hashing algorithm
+if desired by the app developers.
+
+## Alternative Approaches
+
+None.
+
+## Decision
+
+Give app developers a way to provide their own hash function.
+
+## Detailed Design
+
+Add `hashFn HashFn` argument to `NewNode` constructor in `node.go`.
+
+```go
+type HashFn interface {
+    // New returns a new hash.Hash calculating the given hash function.
+    New() hash.Hash
+    // Size returns the length, in bytes, of a digest resulting from the given hash function.
+    Size() int
+}
+
+type TMHash struct {}
+func (TMHash) New() hash.Hash {
+    return tmhash.New()
+}
+func (TMHash) Size(bz []byte) {
+    return tmhash.Size
+}
+
+func DefaultHashFn() HashFn {
+    return TMHash{}
+}
+```
+
+And then use it to calculate a transaction's hash.
```

Comment:
Can you be more specific regarding that?

In the case of `types/Tx`, calculating the hash is hard-coded. How can we parameterize this?

Reply:
In the case we want this change to apply to the RPC endpoints and, possibly, indexer is one thing. In this case we can use any customized hash function.

But if we want to adopt customized hash functions for our types, this is a completely different discussion.

Reply:
> But if we want to adopt customized hash functions for our types, this is a completely different discussion.

I'd like for us to have both discussions.